File: openfeed-frontend/lib/supabase/queries/global_article_metadata_options.ts
import { SupabaseClient } from "@supabase/supabase-js";

import { Database } from "@/lib/supabase/supabase.types";

export type ArticleMetadataField =
  | "topic"
  | "type"
  | "coverage"
  | "duration"
  | "impact";

export interface GlobalArticleMetadataOption {
  field: ArticleMetadataField;
  name: string;
  description: string;
}

export type MetadataOptionsByField = Record<ArticleMetadataField, string[]>;

const METADATA_FIELDS: ArticleMetadataField[] = [
  "topic",
  "type",
  "coverage",
  "duration",
  "impact",
];

export async function getGlobalArticleMetadataOptions(
  supabase: SupabaseClient<Database>,
): Promise<GlobalArticleMetadataOption[]> {
  const { data, error } = await (supabase as any)
    .from("global_article_metadata_options")
    .select("field, name, description")
    .order("field", { ascending: true })
    .order("name", { ascending: true });

  if (error) throw new Error(error.message);
  return ((data as GlobalArticleMetadataOption[] | null) ?? []).filter((row) =>
    METADATA_FIELDS.includes(row.field),
  );
}

export function groupMetadataOptionsByField(
  options: GlobalArticleMetadataOption[],
): MetadataOptionsByField {
  const empty: MetadataOptionsByField = {
    topic: [],
    type: [],
    coverage: [],
    duration: [],
    impact: [],
  };

  return options.reduce((acc, option) => {
    acc[option.field].push(option.name);
    return acc;
  }, empty);
}

File: openfeed-backend/openfeed/db/queries/global_article_metadata_options.py
from collections import defaultdict

from pydantic import BaseModel

from openfeed.db.client import Client
from openfeed.db.utils import paginated_query


class GlobalArticleMetadataOption(BaseModel):
    description: str
    field: str
    name: str

    @property
    def label(self) -> str:
        return self.name


def get_global_article_metadata_options(
    db: Client,
) -> dict[str, list[GlobalArticleMetadataOption]]:
    options = [
        GlobalArticleMetadataOption.model_validate(row)
        for page in paginated_query(
            db,
            "global_article_metadata_options",
            select="field, name, description",
        )
        for row in page
    ]
    grouped_options: dict[str, list[GlobalArticleMetadataOption]] = defaultdict(list)
    for option in sorted(
        options,
        key=lambda option: (
            option.field,
            option.name,
        ),
    ):
        grouped_options[option.field].append(option)
    return dict(grouped_options)


def get_global_topics(db: Client) -> list[GlobalArticleMetadataOption]:
    return get_global_article_metadata_options(db).get("topic", [])
